vsr: op_checkpoint_sync for in-memory checkpoint during state sync

chaitanyabhandari opened the pull request:
Introduce `op_checkpoint_sync`, which accounts for in-memory checkpoint during state sync (similar to `op_prepare_max_sync`). Logic that is exercised via state-sync _must_ use `op_checkpoint_sync`, so we change the following:
* `replace_header` (in the assertion which causes the crash in this seed)
* `op_repair_min` 
*  `checkpoint_id_for_op` (invoked via `op_repair_min`)

This simplifies all assertions during state sync, which were _earlier_ of the format `assert(self.syncing == .updating_checkpoint or ...)` to account for state sync to drop the first condition. For example, we now this:
```C
     assert((self.op < vsr.Checkpoint
                    .checkpoint_after(self.op_checkpoint_sync()) + constants.journal_slot_count));
```

Instead of:
```C
    assert(self.syncing == .updating_checkpoint or
        (self.op < self.op_checkpoint_next() + constants.journal_slot_count));
```

matklad commented:
Very valguely related, we currently have:

```
            switch (self.syncing) {
                .idle => {},
                .canceling_commit, .canceling_grid => {
                    // Ignore further messages until finishing (asynchronous) processing of sync SV.
                    // Notably, this prevents our view from jumping ahead of SV.
                    assert(self.sync_start_view != null);
                    log.warn("{}: on_message: ignoring (syncing)", .{self.log_prefix()});
                    return;
                },
                .updating_checkpoint => {},
            }
```

Would be great to move that over to suspend machinery (it wasn't their once that code was written)

matklad commented:
Thanks so much for the write up here chai, this code is always painful to load into mental instruction cache, so the help is much appreciated!

This indeed goes to the heart of why this is so convoluted: logically, our state is log+checkpoint, and, during sync, we would like to atomically swap it, but, as the saying goes, one does not simply...

The two obstacles for atomicity are:

* any pending state-machine writes from commit we are cancelling need to complete before we change anything
* and then updating checkpoint requires disk write, so it is also async...

Ok, gotta run, will finish comment later, but what I wanted to say is:

* let's keep updates to op and commit_min atomic
* let's explicitly handle checkpoint not being in sync.
* In particular, see how `op_prepare_max_sync` is used
* and perhaps in `on_start_view_set_checkpoint` we need to move

```
            if (self.commit_stage == .checkpoint_superblock) return true;
            if (self.commit_stage == .checkpoint_data) return true;
            if (self.syncing == .updating_checkpoint) return true;
```

cases up? In particular, it feels like if `syncing == .updaing_checkpoing`, we should _never_ return  `false` there?

chaitanyabhandari commented:
Great catch re: https://github.com/tigerbeetle/tigerbeetle/pull/3371#issuecomment-3527848040 @matklad, I will take that up as well!

Also some additional context, which I just dug up... the reason why this failure showed up right now is because https://github.com/tigerbeetle/tigerbeetle/pull/3294 refactored the corresponding assert: https://github.com/tigerbeetle/tigerbeetle/pull/3294/files#diff-e95e4dc6c1aae349e798cfeed22ad01f830cce0e654e6b4b765b3e7b0c677b6aL8266

Specifically, we changed it from `assert(self.syncing == .updating_checkpoint or self.journal.has_header(header));` → ` assert(self.journal.has_header(header));`. We _could_ fix seed by changing the assert back, but I  _think_ this change is a net positive so we could just keep it as is? 🤔

chaitanyabhandari commented:
Reply to https://github.com/tigerbeetle/tigerbeetle/pull/3371#issuecomment-3527951141 ([my earlier comment](https://github.com/tigerbeetle/tigerbeetle/pull/3371#issuecomment-3527961525) was written before I read your comment @matklad):

AH, one of my potential solutions here which I discarded was introducing a `op_checkpoint_sync`, similar to `op_prepare_max_sync`! And also, my other potential solution here (which I _did_ push to an initial version of this PR but reverted later) was to just move `if (self.syncing == .updating_checkpoint) return true;` up so it always returns `true`... I guess I need an "Alternate Approaches" section in the PR description (always admired DJ for doing that) spelling out the additional ways this could've been solved. :) 

The reason why I discarded the _latter_ (moving `if (self.syncing == .updating_checkpoint) return true;` up) was that I felt that we wouldn't _need_ to do that if we just made `on_start_view_set_journal` safe to invoke multiple times while we're in `.updating_checkpoint` status. But if we do want to keep `op` and `commit_min` update atomic, and then maintain the invariant that we only call `on_start_view_set_journal` _once_ while we are updating our checkpoint, we would have to pull this condition up indeed!

Question: I didn't quite understand the _why_ of keeping the update of `op` and `commit_min` atomic... In my mind, it feels like the `op` and `commit_max` are intertwined (and so should be in sync), but `commit_min` _could_ be lagging, no? Also, I am not a 100% sold on my solution, just wanted to understand your thoughts here better. :D

matklad commented:
Argh, gotta run again, but I want to say that I now agree that we _shouldn't_ move `return true;` check up! The idea (which I know recall) is to always keep replica in a valid state, and that means replicas should always be able to act on SV message.

We shouldn't ignore the message just because it might be incosisent with our asserts, rather, we should tighten asserts to say "hey, if this looks inconsistent, it is because of that specific good reason"


>Specifically, we changed it from assert(self.syncing == .updating_checkpoint or self.journal.has_header(header)); →  assert(self.journal.has_header(header));

Ah, I was wondering why we haven't seen this earlier! 


(explaining the atomicity of op+commit_min is still on the liabilities for me)

matklad commented:
>the why of keeping the update of op and commit_min atomic... In my mind, it feels like the op and commit_max are intertwined (and so should be in sync), but commit_min could be lagging, no?

The way I see it, the state we have is the log --- the linked list of prepares. op, commit_min and commit_max are essentially pointers into the log. They aren't just numbers, they point to a very specific prepare and say "hey, this one over here is committed". 

So, whenever we change the entire log, we have to adjust op, commit min, and commit max, to make sure they are still valid, otherwise we might get into a "dangling pointer" kind of situation where, say `commit_min` refers to a different prepare then the one we had in mind originally. 

I am actually surprised that the current diff _doesn't_ trip any asserts in VOPR. I _think_ we can get in a situation where our `op` is on the checkpoint we synced to, while our `commit_min` is on the old checkpoint, so the diff between the two is much larger than journal_slot_count, which feels like an invariant we should be asserting somewhere? 

Though, if it doesn't fail, maybe I am misreading the code here? Will take another look tomorrow!

chaitanyabhandari commented:
> The way I see it, the state we have is the log --- the linked list of prepares. op, commit_min and commit_max are essentially pointers into the log. They aren't just numbers, they point to a very specific prepare and say "hey, this one over here is committed". 
>
> So, whenever we change the entire log, we have to adjust op, commit min, and commit max, to make sure they are still valid, otherwise we might get into a "dangling pointer" kind of situation where, say commit_min refers to a different prepare then the one we had in mind originally.

Ah _great_ point there, that refines my understanding a bit as well. So I was right in saying that commit_min _could_ be lagging, but at the end of the day, it should only be lagging by a finite amount... That is, `commit_min` should be >= `op_checkpoint()`, even if `op` and `commit_max` are far ahead! I don't think that we can assert anything that involves _all three_ (though I may be wrong and I'll think about it more)? 🤔 


> I am actually surprised that the current diff doesn't trip any asserts in VOPR. I think we can get in a situation where our op is on the checkpoint we synced to, while our commit_min is on the old checkpoint, so the diff between the two is much larger than journal_slot_count, which feels like an invariant we should be asserting somewhere?

I believe that could be the reason why we don't fail, we must be asserting the `commit_min` should be >= `op_checkpoint()` , but we don't violate that during this time, because `op_checkpoint()` while we're in `.updating_checkpoint` is still the _old_ checkpoint, so our _old_ commit_min is compatible with that. Also, another point here is that while we're in `updating_checkpoint`, we aren't really doing anything... we short circuit `repair`, `commit_journal`, and `on_prepare` based on that condition!

> We shouldn't ignore the message just because it might be incosisent with our asserts, rather, we should tighten asserts to say "hey, if this looks inconsistent, it is because of that specific good reason"

Also, whole heartedly agre! This is precisely why I was treating the `assert(self.syncing == .updating_checkpoint or self.journal.has_header(header));` →  `assert(self.journal.has_header(header));` change not as a bug, but just an opportunity to improve our code flow, and make `start_view_set_journal` safe to invoke multiple times... I am still open to the idea of keeping the `op`, `commit_min`, and `commit_max` atomic, and making the invocation of `start_view_set_journal` safe by changing the assertion to use `op_checkpoint_sync`. Would love to also hear your thoughts there @matklad. :D 

```C
 if (self.op_checkpoint_sync() < header.op and header.op <= self.commit_min) { 
     if (self.journal.header_with_op(header.op)) |_| { 
         assert(self.journal.has_header(header)); 
     } 
 } 
```

chaitanyabhandari commented:
Took a stab at `op_checkpoint_sync` in https://github.com/tigerbeetle/tigerbeetle/pull/3371/commits/070eae128934a1354b7418382a641cc26e31691a! 

One thing I am not 100% sure about is the naming for `checkpoint_id_for_op`, which now branches on the sync status to decide which checkpoint to look up the `checkpoint_id` from. 
* _Changing_ it to `checkpoint_id_for_op_sync` feels wrong because then we change all the call sites, even the ones that _aren't_ exercised while we are in `updating_checkpoint` status. This would break our current convention, where `op_prepare_max_sync` and `op_checkpoint_sync` are used only in code paths that *can* be exercised during `updating_checkpoint`, while other paths still use `op_prepare_max` and `op_checkpoint`.
* _Adding_ a new `checkpoint_id_for_op_sync` would require ~20 lines of duplicate logic. *Or* the other option is to parametrize `checkpoint_id_for_op` to take in `(self: *const Replica, op: u64, checkpoint: *const vsr.CheckpointState)`, but then we change all call sites (which is not too many, just 5).

My hunch says the latter makes more sense, but would love to hear your thoughts!

matklad commented:
Hm, taking a step back, would restoring the original assertion guard fix the issue?

```diff
diff --git a/src/vsr/replica.zig b/src/vsr/replica.zig
index 26c45c661..ee156986e 100644
--- a/src/vsr/replica.zig
+++ b/src/vsr/replica.zig
@@ -8339,7 +8339,7 @@ pub fn ReplicaType(
 
             if (self.op_checkpoint() < header.op and header.op <= self.commit_min) {
                 if (self.journal.header_with_op(header.op)) |_| {
-                    assert(self.journal.has_header(header));
+                    assert(self.syncing == .updating_checkpoint or self.journal.has_header(header));
                 }
             }
 
``` 

Looking at the code, we actually have a bunch of `assert(self.syncing == .updating_checkpoint or ` guards elsewhere, so maybe we should stick to relaxed invariants? This might not be maximally tight, but it feels significantly simpler!

chaitanyabhandari commented:
@matklad Ah, I had forgotten to update few occurrences yesterday, but introducing `op_checkpoint_sync` actually removes _all_ occurrences of `assert(self.syncing == .updating_checkpoint or`! Updated in https://github.com/tigerbeetle/tigerbeetle/pull/3371/commits/f4b68339d0e041762d404fc3252ddce4cb53a20b. 

To answer your question though, yes, just changing the assert back to `assert(self.syncing == .updating_checkpoint or self.journal.has_header(header));` would fix this issue. :) 

I also realized yesterday that its perhaps incorrect to _not_ have `op_checkpoint_sync`, because we have the following condition in `replace_header`? Without being aware of the in-memory checkpoint during `updating_checkpoint`, `op_repair_min` actually returns an op from the _old_ checkpoint, making this condition trivially true when we're replacing our headers with SV headers from a future checkpoint.
```C
            if (header.op < self.op_repair_min()) return;
```

matklad commented on src/vsr/replica.zig:
```diff
@@ -2852,8 +2853,25 @@ pub fn ReplicaType(
                                 message.header.commit_max,
                                 @src(),
                             );
+
```
Did we intend to add this blank line and the one above? I like how previously the assert on the output of the functional call was directly attached to the call itself. With a blank, it's ambiguous whethere we assert the result of what we did before, or the precondition of what goes after. 

And the 2837 blank feels like it just uses up vertical space without clear benefit to readability? In general, blank lines are as important, if not more important than lines with code, we use them to reveal the structure!

chaitanyabhandari replied:
Ah, both stray and unintentional lines, thank you for pointing them out @matklad! The second one was a remnant from when I had the logic to update commit_min there, but then I decided to move that logic down, _after_ the asserts for `set_op_and_commit_max`.

```C
                self.set_op_and_commit_max(
                    header.op,
                    message.header.commit_max,
                    @src(),
                );
                assert(self.op == header.op);
                assert(self.commit_max >= message.header.commit_max);

                if (self.syncing == .updating_checkpoint) {
```

matklad commented:
```
                if (self.journal.header_with_op(header.op)) |_| {
                    assert(self.syncing == .updating_checkpoint or self.journal.has_header(header));
                }
```

needs update after rebase! Otherwise, LGTM!

chaitanyabhandari committed:
vsr: op_checkpoint_sync for in-memory checkpoint during state sync

chaitanyabhandari commented:
Addressed comments in https://github.com/tigerbeetle/tigerbeetle/pull/3371/commits/12d32ec532e3478cd7def3cae1b64cfbc94ecac2, thanks for the review and great discussions @matklad!
